File: pages/api/balance_sheet.ts
// Next.js API route support: https://nextjs.org/docs/api-routes/introduction
import type { NextApiRequest, NextApiResponse } from "next";

import { SHEET } from "@/config/sheet";

export default function handler(req: NextApiRequest, res: NextApiResponse) {
  if (req.method !== "GET") {
    return res.status(405).end();
  }
  let errorMsg = "";
  try {
    const { provider, company, amount } = req.query;
    if (!provider || typeof provider !== "string") {
      errorMsg = "Invalid Application Provider";
      throw new Error(errorMsg);
    }
    if (!company || typeof company !== "string") {
      errorMsg = "Invalid company";
      throw new Error(errorMsg);
    }
    if (!amount || typeof amount !== "string" || parseInt(amount) <= 0) {
      errorMsg = "Invalid loan amount";
      throw new Error(errorMsg);
    }
    const result = SHEET.filter(
      (sheet) => sheet.company === company && sheet.provider === provider
    );
    if (result.length === 0) {
      errorMsg = "Combination of company and provider does not exist.";
      throw new Error(errorMsg);
    }
    const today = new Date();
    const cutOffDate = new Date(
      today.getFullYear() - 1,
      today.getMonth(),
      today.getDate()
    );
    let resultItem = result[0];
    let preAssessment = 20;
    const filteredData = resultItem.data.filter(
      (item) => new Date(item.year, item.month) >= cutOffDate
    );
    const totalProfitOrLoss = filteredData.reduce((total, item) => {
      return total + item.profitOrLoss;
    }, 0);
    const totalAssetsValue = filteredData.reduce((total, item) => {
      return total + item.assetsValue;
    }, 0);
    resultItem.data = filteredData;

    if (totalProfitOrLoss > 0) {
      preAssessment = 60;
    }
    if (totalProfitOrLoss > 0 && totalAssetsValue / 12 > parseInt(amount)) {
      preAssessment = 100;
    }

    res.status(200).json({
      ...resultItem,
      totalAssetsValue: totalAssetsValue,
      totalProfitOrLoss: totalProfitOrLoss,
      preAssessment: preAssessment,
    });
  } catch (error) {
    return res.status(400).json({ message: errorMsg });
  }
}

File: backend/resources/balance_sheet.py
import json
from datetime import datetime, timedelta

from flask import request
from flask_restful import Resource


class BalanceSheetResource(Resource):
    def get(self):
        params = request.args
        if "provider" not in params:
            return {"message": "Invalid application provider"}, 400
        if "company" not in params:
            return {"message": "Invalid company"}, 400
        if "amount" not in params or int(params["amount"]) <= 0:
            return {"message": "Invalid amount"}, 400
        company, provider, amount = (
            params["company"],
            params["provider"],
            params["amount"],
        )
        total_profit_or_loss = 0
        total_assets_value = 0
        pre_assessment = 20
        now = datetime.now()
        last_12_months = now - timedelta(days=365)
        with open("SHEET.json") as f:
            data = json.load(f)
            for d in data:
                if d["company"] == company and d["provider"] == provider:
                    details = d["data"]
                    for dt in details:
                        if (
                            datetime(dt["year"], dt["month"], now.day + 1)
                            >= last_12_months
                        ):
                            total_assets_value += dt["assetsValue"]
                            total_profit_or_loss += dt["profitOrLoss"]
                    if total_profit_or_loss > 0:
                        pre_assessment = 60
                    if total_profit_or_loss > 0 and total_assets_value / 12 > int(
                        amount
                    ):
                        pre_assessment = 100
                    return {
                        "totalAssetsValue": total_assets_value,
                        "totalProfitOrLoss": total_profit_or_loss,
                        "preAssessment": pre_assessment,
                        "data": d,
                    }, 200
        return {"message": "Combination of company and provider does not exist."}, 400
